add oci pull to images

mac641 opened the pull request:
## Description

### References

- #169
- #154
- https://github.com/metal-stack/pixie/pull/40
- https://github.com/metal-stack/metal-api/pull/635

### TODO

- [x] how are oci credentials supposed to be retrieved?
- [x] add tests

<!--
If possible, please reference other issues or pull requests.

Closes #<the-issue-number-to-close>.

References:

- ...

If not already described in a referenced issue, please describe your PR and the motivation behind it. Just try to make life easy for the reviewers.

Please be aware that the pull request's title will become part of the release notes, so try to make it understandable.
-->

<!--
You maybe want to attach the triage label if you want it to be discussed in the next planning meeting. It might be useful to attend the meeting if you want to emphasize it.

If you would like to add something to the release notes for the next metal-stack release (metal-stack/releases), you can do so by adding SPECIAL SECTIONS (code blocks) in this PR. Please only add a section when this is relevant for the entire project.

You can use the following snippets as an example:

## Release Notes

### Breaking Change

```BREAKING_CHANGE
Description of the breaking change and what an operator needs to do about it.
This section is **not** intended for documentation of internal breaking changes.
Release notes are meant to be read by users and operators of metal-stack, not metal-stack developers.
```

### Required Actions

```ACTIONS_REQUIRED
Description of the required action for operators.
```

### Noteworthy

```NOTEWORTHY
Description of noteworthy release information for the metal-stack project that users or operators should know.
```
-->

mac641 committed:
feat(image): add OciPull based on the example from #169

mac641 commented:
@majst01 okay, so this is what I came up with. Can you do a review, pls?

Besides, I'm wondering if it's suitable to retrieve the oci credentials from env vars. Can you help me out here?

majst01 commented:
> @majst01 okay, so this is what I came up with. Can you do a review, pls?
> 
> Besides, I'm wondering if it's suitable to retrieve the oci credentials from env vars. Can you help me out here?

The OCI creadentials should be passed as optional fields from the pixiecore to the metal-hammer:

https://github.com/metal-stack/metal-hammer/blob/master/cmd/pixie-client.go#L14

Ideally a slice of registryconfig which contains the registry url, the username and the password should come from pixiecore.

mac641 committed:
refactor: use OciConfig in order to retrieve credentials

mac641 commented:
@majst01 does this work? https://github.com/metal-stack/metal-hammer/pull/175/files#diff-cfcd5a78ad9476ee24d7680dd023e71136f329c771d76ae6837ba68320b6379cR37

what does `machine.Allocation.Image.URL` contain? Is it the full oci url or just the image name, e.g. `debian:12`?

majst01 commented:
> @majst01 does this work? https://github.com/metal-stack/metal-hammer/pull/175/files#diff-cfcd5a78ad9476ee24d7680dd023e71136f329c771d76ae6837ba68320b6379cR37
> 
> what does `machine.Allocation.Image.URL` contain? Is it the full oci url or just the image name, e.g. `debian:12`?

It contains the full URL as configured in the image

mac641 committed:
chore: run go mod tidy

mac641 committed:
refactor: add logging to image and format install.go

mac641 committed:
fix: fix issue where RegistryURL was mandatory even though

machine.Allocation.Image.URL can be used without auth

mac641 committed:
chore: apply latest pixie changes where OciConfig.RegistryURL is optional now

mac641 committed:
fix(image): remove optional oci:// prefix before parsing image reference

since it's not supported by name.ParseReference

mac641 committed:
fix(install): add missing check if no OciConfigs have been provided and default to image

mac641 committed:
fix: fix untar destination of oci images and various typos

mac641 committed:
test: add tests to image's oci-puller

mac641 committed:
test(image): implement test for oci image pull from registry with authentication

mac641 committed:
refactor(image): remove external dependency from ghcr in tests

mac641 committed:
fix(image): remove dependency for running tests with superuser permissions

mac641 committed:
ci: separate oci-pull tests from legacy tests

since they're considered integration tests and can't be run inside
a docker container due to testcontainers-go

majst01 commented on cmd/install.go:
```diff
@@ -30,15 +31,34 @@ func (h *hammer) Install(machine *models.V1MachineResponse) (*api.Bootinfo, erro
+				if err != nil {
+					return nil, err
+				}
+			}
```
The ociconfig should be a map from registry -> credentials, then you  can simple get the credentials required for this registry and no need to iterate

majst01 commented on cmd/image/image.go:
```diff
@@ -171,3 +216,71 @@ func (i *Image) download(source, dest string) error {
+
+		target := filepath.Join(dest, hdr.Name)
+
+		fmt.Printf("extracting:%s\n", target)
```
i.log.Debug

majst01 commented on cmd/image/image.go:
```diff
@@ -171,3 +216,71 @@ func (i *Image) download(source, dest string) error {
+		fmt.Printf("extracting:%s\n", target)
+
+		if strings.HasSuffix(target, ".log") {
+			fmt.Printf("skipping:%s\n", target)
```
i.log.Debug

majst01 commented on go.mod:
```diff
@@ -1,55 +1,86 @@
+	github.com/foomo/htpasswd v0.0.0-20200116085101-e3a90e78da9c
+	github.com/metal-stack/metal-lib v0.23.5
 )
 
```
no additional require

majst01 commented on go.mod:
```diff
@@ -1,55 +1,86 @@
 	// keep this until https://github.com/u-root/u-root/pull/3451 is merged and released
 	github.com/u-root/u-root => github.com/majst01/u-root v0.0.0-20250910091544-306665b6f8e8
+
+	google.golang.org/genproto => google.golang.org/genproto v0.0.0-20250826171959-ef028d996bc1
```
why is this required ?

mac641 replied:
I had an import cycle which was fixed by this. If you have any idea on how to solve this differently, I'm happy to implement it.

mac641 committed:
refactor: address changes requested by @majst01

majst01 commented on cmd/image/image.go:
```diff
@@ -171,3 +217,71 @@ func (i *Image) download(source, dest string) error {
+		}
+
+		switch hdr.Typeflag {
+		case tar.TypeDir:
```
Unsure if all cases are covered and or relevant.

Other pitfalls might be:
- files with s-bit set
- file modes of special files like /etc/shadow /etc/passwd etc.

This should be tested, at least manually, better would be with a automated test ?

majst01 commented on cmd/image/image_integration_test.go:
```diff
@@ -0,0 +1,187 @@
+
+func TestOciPull(t *testing.T) {
+	var (
+		assert = assert.New(t)
```
please remove

majst01 commented on cmd/image/image_integration_test.go:
```diff
@@ -0,0 +1,187 @@
+
+		i := NewImage(slog.Default())
+		if err = i.OciPull(ctx, imageRef, mountDir, anonymousUsername, anonymousPassword); err != nil {
+			t.Error(err)
```
require.NoError

majst01 commented on cmd/image/image_integration_test.go:
```diff
@@ -0,0 +1,187 @@
+	var (
+		assert = assert.New(t)
+
+		ctx          = context.Background()
```
```suggestion
		ctx          = t.Context()
```

mac641 committed:
test(image): address changes requested by @majst01

mac641 committed:
refactor(image): add logging to untar's default case in order to detect skipped files

  - reduce code duplication in untar func

mac641 committed:
test(image): add basic working test for untar

mac641 commented:
@majst01 So, I've been adding a couple of tests, mainly targeting s-bits and symlinks but it's impossible to test special files like `/etc/shadow` etc. without root permissions. I was considering an in-memory filesystem like afero as well but this would require dependency injection to work which would end up in a lot of refactoring. I think this is out-of-scope.
What do you think? Shall we skip the special files test? How would you tackle this problem?

mac641 committed:
chore: add latest oci-puller pixie version

mac641 committed:
Merge branch 'master' into oci-puller

mac641 committed:
chore: add latest oci-puller pixie version

mac641 committed:
refactor: log ociConfig for debugging

vknabel commented on cmd/image/image.go:
```diff
@@ -25,7 +34,45 @@ func NewImage(log *slog.Logger) *Image {
+		return fmt.Errorf("parsing image reference: %w", err)
+	}
+
+	// Choose authentication method
```
I think the code is self-explanatory here.

vknabel commented on cmd/image/image.go:
```diff
@@ -25,7 +34,45 @@ func NewImage(log *slog.Logger) *Image {
+	imageRefWithoutOciPrefix := strings.TrimPrefix(imageRef, "oci://")
+
+	i.log.Info("log image ref without oci prefix", "imageRefWithoutOciPrefix", imageRefWithoutOciPrefix)
+	// Parse the image reference (e.g., docker.io/library/alpine:latest)
```
```suggestion
```

I think the comment and the log message aren't needed here.

mac641 replied:
This is just for testing in the mini-lab. I'd rather keep it until this works but you're right, we don't need in the end.

vknabel commented on cmd/image/image.go:
```diff
@@ -25,7 +34,45 @@ func NewImage(log *slog.Logger) *Image {
+		return fmt.Errorf("extracting tar: %w", err)
+	}
+
+	i.log.Info("pull oci image done", "image", imageRef)
```
This log is fine, but the wording sounds a bit off.

vknabel commented on cmd/install.go:
```diff
@@ -29,16 +30,29 @@ func (h *hammer) Install(machine *models.V1MachineResponse) (*api.Bootinfo, erro
-	err = img.NewImage(h.log).Burn(h.chrootPrefix, image, h.osImageDestination)
-	if err != nil {
-		return nil, err
+		h.log.Info("log oci config", "ociConfig", ociConfig)
```
Either remove log or rephrase

mac641 replied:
This is also just for testing in the mini-lab. I'll add a TODO comment two both log statements in order to remove it before merging.

vknabel commented on cmd/image/image_test.go:
```diff
@@ -41,5 +45,91 @@ func TestCheckMD5(t *testing.T) {
+		})
+	}
+
+	// TODO: consider character and block devices -> skipped for now due to root permissions
```
I agree that a potential `afero` refactoring is out of scope for this PR. But I think it would be nice to do (I'm not a maintainer though).
These (and the above FIXME and similar cases) would be really important.

mac641 replied:
@majst01 what are your thoughts about this?

mac641 committed:
refactor: address changes requested by @vknabel
